[quant] Reference option for conv module

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52316 [quant] Reference option for conv module**

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26505642](https://our.internmc.facebook.com/intern/diff/D26505642)

jerryzh168 committed:
[quant] Reference option for conv module

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Reference option for conv module"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Reference option for conv module"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Reference option for conv module"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Reference option for conv module"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Reference option for conv module"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26505642](https://our.internmc.facebook.com/intern/diff/D26505642)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Reference option for conv module"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26505642](https://our.internmc.facebook.com/intern/diff/D26505642)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Reference option for conv module"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26505642](https://our.internmc.facebook.com/intern/diff/D26505642)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Reference option for conv module"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26505642](https://our.internmc.facebook.com/intern/diff/D26505642)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Reference option for conv module"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26505642](https://our.internmc.facebook.com/intern/diff/D26505642)

[ghstack-poisoned]

vkuzo commented on torch/nn/quantized/_reference/modules/conv.py:
```diff
@@ -0,0 +1,136 @@
+from torch.nn.common_types import _size_1_t
+from torch.nn.modules.utils import _single
+
+class _ConvNd(nnq._ConvNd):
```
I don't feel too strongly one way or the other and haven't thought through it in detail, just wondering, do we want reference modules to inherit from quantized modules, or do we want the reverse?  I would have expected the reverse

jerryzh168 replied:
I see, yeah I think it probably makes more sense to do the reverse

jerryzh168 replied:
let me do it in a different PR, we need to change linear as well

vkuzo replied:
sounds good

jerryzh168 committed:
Update on "[quant] Reference option for conv module"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26505642](https://our.internmc.facebook.com/intern/diff/D26505642)

[ghstack-poisoned]

vkuzo commented on torch/nn/quantized/_reference/modules/conv.py:
```diff
@@ -0,0 +1,136 @@
+    """
+    __annotations__ = {"_bias": Optional[torch.Tensor]}
+
+    def _save_to_state_dict(self, destination, prefix, keep_vars):
+        super()._save_to_state_dict(destination, prefix, keep_vars)
+        destination[prefix + '_qweight'] = self._qweight
+        destination[prefix + '_bias'] = self._bias
+
+    def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
```
I didn't check what's currently in the corresponding functions in nnq._ConvNd, but is there a way to reuse these / etc, just to minimize the LOC footprint?  Makes sense if there is a good reason for it.

jerryzh168 replied:
they are not reusable as commented below.

vkuzo commented on torch/nn/quantized/_reference/modules/conv.py:
```diff
@@ -0,0 +1,136 @@
+            state_dict, prefix, local_metadata, False,
+            missing_keys, unexpected_keys, error_msgs)
+
+    def _weight_bias(self):
+        return self._qweight, self._bias
+
+    def set_weight_bias(self, w: torch.Tensor, b: Optional[torch.Tensor]) -> None:
+        self._qweight = w
+        self._bias = b
```
I guess these ones are not reusable because the weight is stored unpacked?

jerryzh168 replied:
yes, that's correct

vkuzo commented on torch/nn/quantized/_reference/modules/conv.py:
```diff
@@ -0,0 +1,136 @@
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x_dequant = x.dequantize()
+        weight_dequant = self._qweight.dequantize()
+        float_result = F.conv1d(
```
if `F.conv1d|F.conv2d|F.conv3d` is stored on the class, can we move `forward` to `_ConvNd` to reuse some more code?

jerryzh168 replied:
currently we have some hacks for conv1d module, but we should be able to merge it after the refactor

vkuzo commented on torch/nn/quantized/_reference/modules/conv.py:
```diff
@@ -0,0 +1,136 @@
+        self._conv1d_dilation = (self.dilation[0],)
+
+class Conv2d(_ConvNd, nnq.Conv2d):
+    def __init__(self, in_channels, out_channels, kernel_size, stride=1,
+                 padding=0, dilation=1, groups=1, bias=True,
+                 padding_mode='zeros'):
+        nnq.Conv2d.__init__(
+            self, in_channels, out_channels, kernel_size, stride, padding, dilation,
+            groups, bias, padding_mode)
```
maybe we can delete this?

jerryzh168 replied:
OK

vkuzo commented on torch/nn/quantized/_reference/modules/conv.py:
```diff
@@ -0,0 +1,136 @@
+        return 'QuantizedConv2d(Reference)'
+
+class Conv3d(_ConvNd, nnq.Conv3d):
+    def __init__(self, in_channels, out_channels, kernel_size, stride=1,
+                 padding=0, dilation=1, groups=1, bias=True,
+                 padding_mode='zeros'):
+        nnq.Conv3d.__init__(
+            self, in_channels, out_channels, kernel_size, stride, padding, dilation,
+            groups, bias, padding_mode)
```
and this?

jerryzh168 replied:
this is multiple inheritance, can we do that?

vkuzo reviewed (approved):
lg, optional comments inline but non-blocking

jerryzh168 committed:
Update on "[quant] Reference option for conv module"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26505642](https://our.internmc.facebook.com/intern/diff/D26505642)

[ghstack-poisoned]
